src/lib.rs

```diff
@@ -891,290 +869,166 @@ impl<L: SynthLanguage> Synthesizer<L> {
-        // Dispose of the runner and only apply unions
-        // to not introduce new (allowed) enodes.
-        log::info!("running allowed rules");
+        // 3. Run allowed rules
+        log::info!("Running allowed rules");
         let runner = self.mk_cvec_less_runner(self.egraph.clone());
-        let rewrites: Vec<&Rewrite<L, SynthAnalysis>> = new_allowed_eqs
-            .values()
-            .chain(old_allowed_eqs.values())
-            .flat_map(|eq| &eq.rewrites)
-            .collect();
-        let (_, found_unions) = self.run_rewrites_with_unions(rewrites, runner);
-        for ids in found_unions.values() {
-            for win in ids.windows(2) {
-                self.egraph.union(win[0], win[1]);
+        let rewrites = allowed.values().flat_map(|eq| &eq.rewrites).collect();
+        let (_, found_unions, _) = self.run_rewrites_with_unions(rewrites, runner);
+        for (canonical_id, other_ids) in found_unions {
+            for id in other_ids {
+                self.egraph.union(canonical_id, id);
             }
         }
-
         self.egraph.rebuild();
 
-        // 2) Lifting rules
-        // This phase denotes newly added allowed nodes in terms
-        // of a forbidden denotation.
-        // It is essential that these rules fully saturate
-        // and that extraneous nodes are kept to a minimum
-        // Keep the runner egraph since it has
-        // the fully denoted forms.
-        // It is possible to learn rules during this phase
-        // so extract potential candidates for
-        // every union.
-        log::info!("running lifting rules");
-        let runner = self.mk_cvec_less_runner(self.egraph.clone());
-        let rewrites: Vec<&Rewrite<L, SynthAnalysis>> = self.lifting_rewrites.iter().collect();
-        let (new_egraph, found_unions) = self.run_rewrites_with_unions(rewrites, runner);
+        // 4. Run lifting rules
+        log::info!("Running lifting rules");
+        let runner = self
+            .mk_cvec_less_runner(self.egraph.clone())
+            .with_iter_limit(usize::MAX)
+            .with_time_limit(Duration::from_secs(1000))
+            .with_node_limit(usize::MAX);
+
+        let rewrites = self.lifting_rewrites.iter().collect();
+        let (new_egraph, found_unions, stop_reason) =
+            self.run_rewrites_with_unions(rewrites, runner);
+        assert!(
+            matches!(stop_reason, StopReason::Saturated),
+            "lifting rules must saturate. Instead, ended due to {:?}",
+            stop_reason
+        );
 
+        let mut candidates: EqualityMap<L> = EqualityMap::default();
         for ids in found_unions.values() {
-            for win in ids.windows(2) {
-                if self.params.keep_all
-                    || (self.egraph[win[0]].nodes.iter().any(L::is_extractable)
-                        && self.egraph[win[1]].nodes.iter().any(L::is_extractable))
-                {
+            for id1 in ids.clone() {
```

Comment:
Is there a reason for this method of iteration rather than a sliding window of 2?

Reply:
I think a sliding window is better. This is causing the performance blow up. Is there some factor I'm missing?

Reply:
There was a discussion about this in [this slack thread](https://uw-cse.slack.com/archives/C01749FQE4U/p1662153465032329)
the gist is- by using a sliding window and unioning as you go, you only see a subset of potential candidates in an arbitrary/unpredictable order